src/shared/bus-polkit.c

```diff
@@ -379,6 +473,20 @@ int bus_verify_polkit_async(
         if (r < 0)
                 return r;
 
+        if (!qs) {
```

Comment:
Is this condition necessary?

Reply:
It is. On the first call of this function (for given `call`), `qs` will be `NULL`.

Reply:
I mean, here `qs` is always `NULL`, no?

Reply:
No. When a dbus method needs to auth. for a polkit action, it calls `bus_verify_polkit_async()`. The function creates an `AsyncPolkitQueries` instance (and stores it into `registry`), an `AsyncPolkitQuery` instance, and issues an async call to polkit. The dbus method then returns immediately after returning from `bus_verify_polkit_async()`. When the async call finishes, `async_polkit_callback()` gets called, which stores the reply in its associated `AsyncPolkitQuery` object and reschedules the same bus message for processing. This time, when `bus_verify_polkit_async()` is called, it finds that a query for this `call` + `action` combination is already present in `registry`, hence it just processes the reply and returns. If the method then needs to auth. for _another_ action, it calls `bus_verify_polkit_async()` again. This time, no existing query is found in the `registry` (it's a different action than previously), but the `AsyncPolkitQueries` is (as we're still processing the same bus message), hence `qs` is set. The function then proceeds with creating another query (so `qs` contains two items now, each for a different action) and issues another async call to polkit. The dbus method again returns and is later rescheduled by `async_polkit_callback()`. And so on...

I hope this is at least a bit understandable...

(Btw, if this sounds complicated, it's because it is :-) But most of the complexity is due to the async. nature of this and has already been there before.)